presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkNativeTaskExecutorFactory.java

```diff
@@ -425,74 +422,73 @@ public boolean hasNext()
-         *          once the task has terminated
-         *
-         *      For a Result Task, this function will return all the pages and Optional.empty()
-         *          once all the pages have been read and the task has been terminates
-         *
-         *  2) Exception handling
-         *      when there are no pages available, the function checks if the task has finished
-         *      with exceptions and throws the appropriate exception back to spark's RDD processing
-         *      layer
-         *
-         *  3) Statistics collection
-         *      For both, when the task finished successfully or with exception, it tries to collect
-         *      statistics if TaskInfo object is available
+        /**
+         * This function returns the next available page fetched from CPP process
+         * <p>
+         * Has 3 main responsibilities
+         * 1)  wait-for-pages-or-completion
+         * <p>
+         * The thread running this method will wait until either of the 3 conditions happen
+         * *      1. We get a page
+         * *      2. Task has finished successfully
+         * *      3. Task has finished with error
+         * <p>
+         * For ShuffleMap Task, as of now, the CPP process returns no pages.
+         * So the thread will be in WAITING state till the CPP task is done and returns an Optional.empty()
+         * once the task has terminated
+         * <p>
+         * For a Result Task, this function will return pages retrieved from CPP side once we got them.
+         * Once all the pages have been read and the task has been terminates
+         * <p>
+         * 2) Exception handling
+         * The function also checks if the task has finished
+         * with exceptions and throws the appropriate exception back to spark's RDD processing
+         * layer
+         * <p>
+         * 3) Statistics collection
+         * For both, when the task finished successfully or with exception, it tries to collect
+         * statistics if TaskInfo object is available
          *
          * @return Optional<SerializedPage> outputPage
          */
         private Optional<SerializedPage> computeNext()
         {
-            // A while(true) loop is not desirable, but in this case we cannot avoid
-            // it because of Spark'sRDD contract, which is that this iterator either
-            // returns data or is complete. It CANNOT return null.
-            // While the remote task is still running and there is no output pages,
-            // we need to simulate a busy-loop to avoid returning null.
-            while (true) {
-                try {
-                    // For ShuffleMap Task, this will always return Optional.empty()
-                    Optional<SerializedPage> pageOptional = nativeExecutionTask.pollResult();
-
-                    if (pageOptional.isPresent()) {
-                        return pageOptional;
+            try {
```

Comment:
I wonder if we can simplify this code by writing 2 different iterators - 1. shuffleMap task and 2. for result task, 
Reason being 
- Behavior of Shuffle Map task is fundamentally different and is more straight forward (wait for `CompletableFuture` returned by `NativeExecutionTask`) Most of the stage in all of our prod workload would use the ShuffleMap iterator as that is the 99% use-case. 

I am worried the multiple `wait()/notifyAll()` pattern here might be hiding behaviors that are even harder to spot than the current busy-wait bug

Reply:
Thanks for the reviews @shrinidhijoshi. 

Technically waiting for `CompletableFuture` (`CompletableFuture::get()`) is also a busy-waiting which is a not good idea to have it on the main thread. 

On the other hand, for our scenarios - asynchronous communication between different threads/processes, I believe the `wait()/notifyAll()` is the right java primitives to use if we can't use more built-in high level blocks (e.g. blockingQueue etc.)